---
title: "Transcriptomics Tutorial"
author: Wade Boohar, Kayla Xu
date: 09/24/23
updated: 09/25/23
---

```{r setup}
    knitr::opts_knit$set(root.dir = normalizePath("/PATH/TO/analysis_data")) 
```

```{r Package Setup}
# install, only if needed, and load packages (SummarizedExperiment and TCGAbiolinks)
library(TCGAbiolinks)
library(SummarizedExperiment)
```

Load in the RNAseq data

```{r}
rna_query <- GDCquery(project ="TCGA-BRCA",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")

rna_se <- GDCprepare(rna_query)

```


###### (1) Summarized Experiment Basics

Let's take a look at rna_se (RNA Summarized Experiment). Call head() and str(). Also explore the data frame by double clicking on it from the Environment panel (or calling "View(rna_se)").

```{r Exercise 1.1}
# write code here

```

Use the dim() function to find the dimensions of the colData, rowRanges@elementMetadata, and unstranded dataframes. Record them below:

  ~ colData (rows, cols): 
  ~ rowRanges@elementMetadata: 
  ~ unstranded: 

```{r Exercise 1.2}
# write code here

```

Run the following code and fill out a description of the rows/columns of our data frames. You can use this as a cheat sheet moving forward.
  ~ colData
        - Rows: 
        - Columns: 
  ~ rowRanges@elementMetadata
        - Rows: 
        - Columns:
  ~ unstranded (in assays)
        - Rows: 
        - Columns: 
  
```{r Exercise 1.3}
# colData: 


# rowRanges:

# what are the rows... Hint: Do the rows match a dimension of another dataframe?

# for unstranded, don't call anything, just use your logic from earlier!
```

###### (2) ColData

Using the nested dataframe colData in rna_se, view the first 15 values in the age_at_index column.

```{r Exercise 2.1}
# write code here

```

Call summary on the age_at_index column. Record the output below:
    Min.: 
    1st Qu.: 
    Median: 
    Mean: 
    3rd Qu.: 
    Max.: 
    NA's: 
    
```{r Exercise 2.2}
# write code here

```

Create a data frame called rna_clinical that contains data from colData.

```{r Exercise 2.3}



```

Use the as.data.frame() function to change this from an S4 object to a data frame object.

```{r Exercise 2.4}



```

Three columns in rna_clinical contain nested lists. To save the dataframe as a csv later, we need to first remove those columns. Using boolean masks, subset out treatments, primary_site, and disease_type

```{r Exercise 2.5}



```


###### (3) RowRanges

Call head() on rowRanges@elementMetadata:

```{r Exercise 3.1}
# write code here


```

Create a data frame called rna_genes that contains data from elementMetadata.

```{r Exercise 3.2}


```

Use the as.data.frame() function to change this from an S4 object to a data frame object

```{r Exercise 3.3}


```

###### (4) Assays

What are the names of the 6 assays?

```{r Exercise 4.1}
# access the assays here


```

Create a new data frame called "rna_counts" that contains the data from unstranded. Recall the patient with NA age data? We need to exclude that patient. We can use the same mask to do this.

```{r Exercise 4.2}


```

Fetch the data for the first 5 patients and the first 15 rows.

```{r Exercise 4.3}

```


###### (5) Cleaning up your data frames

Call rownames() and colnames() on rna_clinical.

```{r Exercise 5.1}


```

Just like we did to "clinical", we are going to create a column in rna_clinical for age_category. Our cutoff will be "young" as <= 58 y/o, and "old" as > 58 y/o.

There are three age columns in rna_clinical. We'll use the column age_at_index, because it has the fewest number of NA values (only 1).

Age columns:
  rna_clinical$age_at_diagnosis (16 NAs) (in days, not years)
  rna_clinical$age_at_index (1 NA) 
  rna_clinical$paper_age_at_initial_pathologic_diagnosis (19 NAs)
  

Use the information in age_at_index with an ifelse() statement to create a new column in rna_clinical called age_category. Remember ifelse(expression, x, y), where x happens if the expression is true, and y happens if it is false.

```{r Exercise 5.4}



```

----

Let's move on to rna_genes. Call rownames() and colnames() on rna_genes.

```{r Exercise 5.5}



```

As you can see, both the row names and column names are informative.
----

Now that rna_genes has informative row/column names, let's move on to rna_counts. Call rownames() and colnames() on rna_counts.

```{r Exercise 5.7}



```

Uh oh! Looks like neither the row nor column names are very informative. Given what we did to the other data frames, add column and row names to rna_counts. Name the columns based on the patient barcodes (ie the row names for rna_clinical), and name the rows based on the gene id's (ie the column names for rna_gene).

```{r Exercise 5.8}



```

Now, if we look at the data, we'll see the row names set as the gene IDs and the column names as the patient sample barcodes.
```{r}

```

----

The last step in cleanup is to subset out any samples that are of a normal tissue type. Call unique() on the definition column of rna_clinical.

```{r Exercise 5.9}
# write code here]


```

Notice that there are three sample types: "Solid Tissue Normal", "Primary solid Tumor", "Additional Metastatic", and "Metastatic". For now, we are simply going to get rid of the "Solid Tissue Normal" samples, but keep this in mind for your midterm and final projects as these samples could provide you with a good control!

Create a mask that calls "Solid Tissue Normal" FALSE, and the other two sample types TRUE.

```{r Exercise 5.10}
# write code here




```

You will need to apply this mask to two of our three data frames (rna_clinical, rna_genes, rna_counts). Think about which df's involve samples, and apply the mask to these data frames to subset out the normal samples. You may overwrite the data frames in this case.

```{r Exercise 5.11}
# write code here



```

  
Since there still is one patients who has "Not Reported" for vital status in our dataframes, we need to remove the patient's information form both rna_clinical and rna_counts. Use the is.na() function to accomplish this here:

```{r}
 

```

Great! Now all of our data frames are ready for exploration and analysis.

  
###### (6) Saving data as csv

Loading data from TCGA can be time consuming. It's much faster to access csv files.

Save rna_clinical to your local machine

```{r Exercise 2.6}
write.csv(rna_clinical, "../skcm_rna_clinical_data.csv", row.names = TRUE)
```


Save rna_genes to your local machine.

```{r Exercise 3.4}
write.csv(rna_genes, "../skcm_rna_gene_data.csv", row.names = FALSE)
```


Save rna_counts to your local machine (warning: this will likely take a while!)

```{r Exercise 4.4}
write.csv(rna_counts, "../skcm_rna_count_data.csv", row.names = TRUE)
```


###### (6) Preparing for plotting

Make a column in rna_clinical called five_yr_surv that contains TRUEs and FALSEs for whether a patient survived 5 years. Use days_to_death, or (if days_to_death is NULL), days_to_last_follow_up. Note: 1 year = 365.25 days.

```{r Exercise 6.1}
# fill in the nested ifelse to capture TRUE/FALSE values for whether a patient survived 5 years according to their days_to_death 

five_yr_death <- ifelse()

# create your column from your finalized boolean vector (there should be only T/F, no NAs) 
rna_clinical$five_year_surv <- five_yr_death

```

----

Recall the %in% function that allows us to check whether a value is present in a vector. Ex:

```{r Exercise 6.2}
fruit_list <- c("peach", "pomegranate", "cherry", "apricot", "kiwi")
"apple" %in% fruit_list # output: FALSE
"kiwi" %in% fruit_list # output: TRUE
```

Using your knowledge from the literature presentations, Google, etc., pick two genes that are important in skin cancer. List them below.
    ~ Gene 1:
    ~ Gene 2: 
    
Using the %in% function, check that these two genes are in the rna_genes data. If they aren't pick a new gene. Hint: think about which column to check in.

```{r Exercise 6.3}
# write code here


```

Create a mask for geneA and geneB where that gene is TRUE and all others are FALSE.

NOTE: for the rest of this notebook, whenever you see "geneA" or "geneB", change that to the name of your gene(s) for easier reference later on.

```{r Exercise 6.4}
# write code here

geneA_mask <-  # create mask based on the gene name (your mask should be true when the name is 'geneA' and false for any other gene name)

geneB_mask <-  #FILL IN HERE
```

We can use this mask for a few things. First, let's use it to grab the ensembl IDs of our genes.

```{r Exercise 6.5}
geneA_ensembl <-  #FILL IN HERE # think about which column we should apply out mask to
  
geneB_ensembl <-  #FILL IN HERE
```

We can also use the mask to create a list that contains counts for just a single gene. (HINT: this means subsetting for just the geneA row, with all columns included)

```{r Exercise 6.6}
geneA_counts <- unlist(#FILL IN HERE) # unlist() is used to reduce a nested dimension (list within a list) so we can call summary on geneA_counts

geneB_counts <- unlist(#FILL IN HERE)
```

Call summary() on geneA_counts and geneB_counts. Record the output below:

  Gene A
    ~ Min.:274
    ~ 1st Qu.: 2883
    ~ Median: 4446
    ~ Mean: 5004
    ~ 3rd Qu.: 6495
    ~ Max.: 25897
    
  Gene B
  ~ Min.: 205
  ~ 1st Qu.: 11831
  ~ Median: 20086
  ~ Mean: 50336
  ~ 3rd Qu.: 31914
  ~ Max.: 995067

```{r Exercise 6.7}
# write code here

```

###### (7) Plotting

Recall that plot() creates a scatterplot. Use this function to plot geneA vs geneB.

```{r Exercise 7.1}
#par(mar=c(1,1,1,1)) # run this line of code if your plot doesn't appear formatted correctly (can't see labels)
plot(geneA_counts,
     geneB_counts,
     xlab = "x label here",
     ylab = "y label here",
     main = "title"
)
```

Do you notice any patterns? One issue with plotting gene expression counts is that un-normalized expression tends to cluster near the bottom left corner of the scatter plot. A common technique used to better visualize the correlation between the expression of genes is to plot the log of the expressions instead.


```{r}
plot(log10(geneA_counts),
     log10(geneB_counts),
     xlab = "log of Gene A",
     ylab = "log of Gene B",
     main = "title"
)
```


Now, do you notice any patterns? Is there a relationship between the two genes that you chose? 

# Write a few sentence making observations about your plot, noting the significance of what this means, and any key takeaways:






----

Let's visualize these two genes again, this time using box plots. To get both geneA_counts and geneB_counts on one plot, make a data frame using data.frame() that contains those lists as columns. Again, if the distribution of the expression counts is too skewed, scale/normalize the data by plotting the log of the values instead.

```{r Exercise 7.2}
geneAB_counts <- data.frame(#FILL IN HERE)  # create data frame with geneA_counts and geneB_counts

par(mar=c(1,1,1,1))
boxplot(geneAB_counts,
        xlab = "x label",
        ylab = "ylabel")
```

# Write a few sentence making observations about your plot, noting the significance of what this means, and any key takeaways:




----

For the next plots, choose either gene A or gene B to continue with. We will be making a box plot with two boxes, one for patients who survived 5 years, and another for patients who did not.

```{r Exercise 7.3}
#par(mar=c(1,1,1,1))
boxplot(y_axis_data ~ x_axis_data, # think about what data (gene counts or the 5 yr survival column) should be on the x vs y axes
        xlab = "x axis label",
        ylab = "y axis label")

dev.off()
```

Does there appear to be a relationship between expression of your gene and five year survival? 

# Write a few sentence making observations about your plot, noting the significance of what this means, and any key takeaways:

----

Create two more gene_count lists, for MKI67 and PTGS2.

```{r Exercise 7.4}
MKI67_counts <- unlist(#FILL IN HERE)
PTGS2_counts <- unlist(#FILL IN HERE)
```

Create a Draftsman plot by filling in the code below.

```{r Exercise 7.5}
# fill in with your 4 individual gene count lists and their names
geneABCD_counts <- data.frame(#put all your list of genes here)
colnames(geneABCD_counts) <- c(#gene list 1, # gene list 2, # gene list 3, # gene list 4)

# the following three lines set up a clustering color scheme for our plot
# choose a column from rna_clinical to cluster based on (ex: age_category, five_yr_surv, AJCC staging, etc.)
cols <- character(nrow(rna_clinical)) 
cols[rna_clinical$COLUMN == VALUE] <- "blue" # fill in column and value for as many unique values are in your column (ex: 2 for age_category, "young"/"old")
cols[rna_clinical$COLUMN == VALUE] <- "red"

# run this line to create your plot!
pairs(geneABCD_counts, col = cols, lower.panel=NULL)
```

Do there appear to be any relationships between the genes in question? Do you see any clear clustering based on your chosen clinical variables? 

#Write a few sentence making observations about your plot, noting the significance of what this means, and any key takeaways:

